Switch all RNG code to the built-in PRNG

sipa opened the pull request:
This does not remove OpenSSL, but makes our own PRNG the 'main' one; for GetStrongRandBytes, the OpenSSL RNG is still used (indirectly, by feeding its output into our PRNG state).

It includes a few policy changes (regarding what entropy is seeded when).

Before this PR:
* GetRand*:
  * OpenSSL
* GetStrongRand*:
  * CPU cycle counter
  * Perfmon data (on Windows, once 10 min)
  * /dev/urandom (or equivalent)
  * rdrand (if available)
* From scheduler when idle:
  * CPU cycle counter before and after 1ms sleep
* At startup:
  * CPU cycle counter before and after 1ms sleep

After this PR:
* GetRand*:
  * Stack pointer (which indirectly identifies thread and some call stack information)
  * rdrand (if available)
  * CPU cycle counter
* GetStrongRand*:
  * Stack pointer (which indirectly identifies thread and some call stack information)
  * rdrand (if available)
  * CPU cycle counter
  * /dev/urandom (or equivalent)
  * OpenSSL
  * CPU cycle counter again
* From scheduler when idle:
  * Stack pointer (which indirectly identifies thread and some call stack information)
  * rdrand (if available)
  * CPU cycle counter before and after 1ms sleep
  * Perfmon data (on Windows, once every 10 min)
* At startup:
  * Stack pointer (which indirectly identifies thread and some call stack information)
  * rdrand (if available)
  * CPU cycle counter
  * /dev/urandom (or equivalent)
  * OpenSSL
  * CPU cycle counter again
  * Perfmon data (on Windows, once every 10 min)

The interface of random.h is also simplified, and documentation is added.

This implements most of #14623.

sipa committed:
Switch all RNG code to the built-in PRNG.

It includes the following policy changes:
* All GetRand* functions seed the stack pointer and rdrand result
  (in addition to the performance counter)
* The periodic entropy added by the idle scheduler now seeds stack pointer,
  rdrand and perfmon data (once every 10 minutes) in addition to
  just a sleep timing.
* The entropy added when calling GetStrongRandBytes no longer includes
  the once-per-10-minutes perfmon data on windows (it is moved to the
  idle scheduler instead, where latency matters less).

Other changes:
* OpenSSL is no longer seeded directly anywhere. Instead, any generated
  randomness through our own RNG is fed back to OpenSSL (after an
  additional hashing step to prevent leaking our RNG state).
* Seeding that was previously done directly in RandAddSeedSleep is now
  moved to SeedSleep(), which is indirectly invoked through ProcRand
  from RandAddSeedSleep.
* Seeding that was previously done directly in GetStrongRandBytes()
  is now moved to SeedSlow(), which is indirectly invoked through
  ProcRand from GetStrongRandBytes().

fanquake commented:
Compile failure on macOS (10.14.2):
```
./autogen.sh
./configure
make check

  CXX      libbitcoin_util_a-logging.o
  CXX      libbitcoin_util_a-random.o
random.cpp:394:41: error: expected ';' after top level declarator
static unsigned char rng_state[32] = {0} GUARDED_BY(cs_rng_state);
                                        ^
                                        ;
random.cpp:394:42: warning: declaration does not declare anything [-Wmissing-declarations]
static unsigned char rng_state[32] = {0} GUARDED_BY(cs_rng_state);
                                         ^
./threadsafety.h:18:23: note: expanded from macro 'GUARDED_BY'
#define GUARDED_BY(x) __attribute__((guarded_by(x)))
                      ^
random.cpp:395:32: error: expected ';' after top level declarator
static uint64_t rng_counter = 0 GUARDED_BY(cs_rng_state);
                               ^
                               ;
random.cpp:395:33: warning: declaration does not declare anything [-Wmissing-declarations]
static uint64_t rng_counter = 0 GUARDED_BY(cs_rng_state);
                                ^
./threadsafety.h:18:23: note: expanded from macro 'GUARDED_BY'
#define GUARDED_BY(x) __attribute__((guarded_by(x)))
                      ^
2 warnings and 2 errors generated.
make[2]: *** [libbitcoin_util_a-random.o] Error 1
make[1]: *** [check-recursive] Error 1
make: *** [check-recursive] Error 1
```

gmaxwell commented:
Might be desirable to stick the openssl stuff behind a define already,  I expect we'll be able to ship 0.18 without linking bitcoind to openssl.

sipa commented:
@fanquake Hopefully fixed.

@gmaxwell There are probably a few entropy sources we want to add first, I think.

gmaxwell commented:
@sipa I am eager to add entropy sources!  (but I meant behind a ifdef that is currently on... just since I think you're touching the only remaining callsites)

practicalswift commented on src/random.cpp:
```diff
@@ -272,71 +301,137 @@ void GetOSRand(unsigned char *ent32)
-    hasher.Write(buf, 32);
+    std::call_once(g_rng_initialized, [&](){
+        // The first invocation of the RNG always gathers all we have.
+        level = RNGLevel::STARTUP;
```
Very minor nit: Likely a matter of taste, but personally I try to avoid parameter reassignment. Would it be possible to avoid parameter re-assignment and make the parameter `level` const? Feel free to ignore this suggestion.

sipa commented on src/random.cpp:
```diff
@@ -272,71 +301,137 @@ void GetOSRand(unsigned char *ent32)
+    hasher.Write((const unsigned char*)&perfcounter, sizeof(perfcounter));
+
+    // Windows performance monitor data (once every 10 minutes)
+    RandAddSeedPerfmon(hasher);
```
No need for. Perhaps we'll extend the function to do something useful on other platforms too.

sipa commented on src/random.cpp:
```diff
@@ -129,18 +135,41 @@ static bool GetHWRand(unsigned char* ent32) {
 }
 
-static void RandAddSeedPerfmon()
+static void RandAddSeedPerfmon(CSHA512& hasher)
```
I'd rather not. Limiting the ifdefs to just one place is preferable imo.

sipa commented on src/random.cpp:
```diff
@@ -448,10 +557,6 @@ bool Random_SanityCheck()
-    RAND_add((const unsigned char*)&start, sizeof(start), 1);
-    RAND_add((const unsigned char*)&stop, sizeof(stop), 1);
-
     return true;
```
I'm not touching this code.

gmaxwell commented on src/random.cpp:
```diff
@@ -135,6 +135,39 @@ static bool GetHWRand(unsigned char* ent32) {
+            local_hasher.Write(buffer, sizeof(buffer));
+        }
+        // Write benchmark data (into output hasher)
+        int64_t perf = GetPerformanceCounter();
```
SeedTimestamp() ?

gmaxwell replied:
Maybe not because I think you should change the strengthening function to 4-wide sha256d64 in order to increase the "work" done per unit time, and seed timestamp takes a hasher.

sipa replied:
Done (using `SeedTimestamp()`).

Leaving this for now; switching to SHA256-based code requires some refactoring to make sure there are no initialization order issues.

gmaxwell commented on src/random.cpp:
```diff
@@ -272,71 +268,116 @@ void GetOSRand(unsigned char *ent32)
-    memory_cleanse(buf, 64);
+    // High-precision timestamp after sleeping
+    int64_t perfcounter = GetPerformanceCounter();
+    hasher.Write((const unsigned char*)&perfcounter, sizeof(perfcounter));
```
SeedTimestamp()

gmaxwell commented on src/random.cpp:
```diff
@@ -326,6 +359,19 @@ static void SeedSlow(CSHA512& hasher)
+    if (current_time > last_strengthen + 60000) {
+        // Write randomness from the actual RNG in the state, so it goes through the strengthening as well.
+        unsigned char buffer[32];
+        GetRandBytes(buffer, 32);
```
I'd kinda feel better about this if it just directly took the cs_rng_state lock and read the state here instead of calling GetRandBytes... at least then the code won't mysteriously crash if someone upgrades GetRandBytes to use ::Slow.

sipa replied:
I've abstracted the update-the-global-RNG-and-extract-entropy into a separate function, which is invoked from both this seeder and RandProc.

sipa committed:
Sprinkle some sweet noexcepts over the RNG code

sipa commented on src/random.cpp:
```diff
@@ -272,71 +301,137 @@ void GetOSRand(unsigned char *ent32)
+    STARTUP, //!< Called only at startup.
+};
+
+static void ProcRand(unsigned char* out, int num, RNGLevel level)
```
It does throw when threads get interrupted apparently (as it has a sleep inside).

practicalswift replied:
Uh, ok. Thanks!

sipa committed:
Abstract out seeding/extracting entropy into RNGState::MixExtract

sipa committed:
Remove hwrand_initialized.

 All access to hwrand is now gated by GetRNGState, which initializes the hwrand code.

sipa committed:
Automatically initialize RNG on first use.

sipa committed:
Use FRC::randbytes instead of reading >32 bytes from RNG

There was only one place in the codebase where we're directly reading >32 bytes from
the RNG. One possibility would be to make the built-in RNG support large reads, but
using FastRandomContext lets us reuse code better.

There is no change in behavior here, because the FastRandomContext constructor
uses GetRandBytes internally.

sipa committed:
Add thread safety annotations to RNG state

sipa commented on src/random.cpp:
```diff
@@ -345,12 +440,26 @@ void GetStrongRandBytes(unsigned char* out, int num)
+    hasher.Reset();
 }
 
+void GetRandBytes(unsigned char* out, int num) { ProcRand(out, num, RNGLevel::FAST); }
```
Done.

sipa commented on src/random.h:
```diff
@@ -13,33 +13,45 @@
+ * Generate random data via the internal PRNG.
+ *
+ * These functions are designed to be fast (sub microsecond), but do not necessarily
+ * meaninfully add entropy to the PRNG state.
```
Fixed.

sipa commented:
Made a number of improvements, and split the history out into hopefully more self-contained commits.

ken2812221 commented:
In constructor of CTxMemPool require to call random functions, but the constructor of the mutex has not being called yet.
https://github.com/bitcoin/bitcoin/blob/e7b88ecbc920321290941bc68e4a71634889c3cb/src/validation.cpp#L244
```
 	test_bitcoin.exe!SeedStartup(CSHA512 & hasher={...}) 第 417 行	C++
 	test_bitcoin.exe!ProcRand(unsigned char * out=0x00000031da8febb8, int num=8, RNGLevel level=STARTUP) 第 462 行	C++
 	test_bitcoin.exe!GetRandBytes(unsigned char * buf=0x00000031da8febb8, int num=8) 第 476 行	C++
 	test_bitcoin.exe!GetRand(unsigned __int64 nMax=18446744073709551615) 第 491 行	C++
 	test_bitcoin.exe!SaltedTxidHasher::SaltedTxidHasher() 第 1093 行	C++
 	test_bitcoin.exe!boost::tuples::cons<SaltedTxidHasher,boost::tuples::cons<std::equal_to<uint256>,boost::tuples::null_type> >::cons<SaltedTxidHasher,boost::tuples::cons<std::equal_to<uint256>,boost::tuples::null_type> >() 第 284 行	C++
 	test_bitcoin.exe!boost::tuples::cons<mempoolentry_txid,boost::tuples::cons<SaltedTxidHasher,boost::tuples::cons<std::equal_to<uint256>,boost::tuples::null_type> > >::cons<mempoolentry_txid,boost::tuples::cons<SaltedTxidHasher,boost::tuples::cons<std::equal_to<uint256>,boost::tuples::null_type> > >() 第 284 行	C++
 	test_bitcoin.exe!boost::tuples::cons<unsigned __int64,boost::tuples::cons<mempoolentry_txid,boost::tuples::cons<SaltedTxidHasher,boost::tuples::cons<std::equal_to<uint256>,boost::tuples::null_type> > > >::cons<unsigned __int64,boost::tuples::cons<mempoolentry_txid,boost::tuples::cons<SaltedTxidHasher,boost::tuples::cons<std::equal_to<uint256>,boost::tuples::null_type> > > >() 第 284 行	C++
 	test_bitcoin.exe!boost::tuples::tuple<unsigned __int64,mempoolentry_txid,SaltedTxidHasher,std::equal_to<uint256>,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>::tuple<unsigned __int64,mempoolentry_txid,SaltedTxidHasher,std::equal_to<uint256>,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>() 第 485 行	C++
 	test_bitcoin.exe!boost::tuples::cons<boost::tuples::tuple<unsigned __int64,mempoolentry_txid,SaltedTxidHasher,std::equal_to<uint256>,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>,boost::tuples::cons<boost::tuples::tuple<boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByDescendantScore,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>,boost::tuples::cons<boost::tuples::tuple<boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByEntryTime,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>,boost::tuples::cons<boost::tuples::tuple<boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByAncestorFee,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>,boost::tuples::null_type> > > >::cons<boost::tuples::tuple<unsigned __int64,mempoolentry_txid,SaltedTxidHasher,std::equal_to<uint256>,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>,boost::tuples::cons<boost::tuples::tuple<boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByDescendantScore,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>,boost::tuples::cons<boost::tuples::tuple<boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByEntryTime,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>,boost::tuples::cons<boost::tuples::tuple<boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByAncestorFee,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>,boost::tuples::null_type> > > >() 第 284 行	C++
 	test_bitcoin.exe!boost::multi_index::multi_index_container<CTxMemPoolEntry,boost::multi_index::indexed_by<boost::multi_index::hashed_unique<mempoolentry_txid,SaltedTxidHasher,boost::mpl::na,boost::mpl::na>,boost::multi_index::ordered_non_unique<boost::multi_index::tag<descendant_score,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByDescendantScore>,boost::multi_index::ordered_non_unique<boost::multi_index::tag<entry_time,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByEntryTime>,boost::multi_index::ordered_non_unique<boost::multi_index::tag<ancestor_score,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByAncestorFee>,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,std::allocator<CTxMemPoolEntry> >::multi_index_container<CTxMemPoolEntry,boost::multi_index::indexed_by<boost::multi_index::hashed_unique<mempoolentry_txid,SaltedTxidHasher,boost::mpl::na,boost::mpl::na>,boost::multi_index::ordered_non_unique<boost::multi_index::tag<descendant_score,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByDescendantScore>,boost::multi_index::ordered_non_unique<boost::multi_index::tag<entry_time,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByEntryTime>,boost::multi_index::ordered_non_unique<boost::multi_index::tag<ancestor_score,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByAncestorFee>,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,std::allocator<CTxMemPoolEntry> >() 第 177 行	C++
 	test_bitcoin.exe!CTxMemPool::CTxMemPool(CBlockPolicyEstimator * estimator=0x00007ff65afbfe60) 第 328 行	C++
>	test_bitcoin.exe!`dynamic initializer for 'mempool''() 第 244 行	C++
 	test_bitcoin.exe!_initterm(void(*)() * first=0x00007ff65a49c000, void(*)() * last=0x00007ff65a49f848) 第 22 行	C++
 	[外部程式碼]

sipa commented:
@ken2812221 That was helpful, thanks! I think I've fixed it, but the same error still appears. Can you check what line number it's on now?

ken2812221 commented:
```
 	test_bitcoin.exe!std::_Load_relaxed_4(volatile unsigned long * _Tgt=0x0000000000000074) 第 1338 行	C++
 	test_bitcoin.exe!std::_Atomic_load_4(volatile unsigned long * _Tgt=0x0000000000000074, std::memory_order _Order=memory_order_relaxed) 第 1357 行	C++
 	test_bitcoin.exe!std::atomic_load_explicit(const std::_Atomic_uint * _Atom=0x0000000000000074, std::memory_order _Order=memory_order_relaxed) 第 495 行	C++
 	test_bitcoin.exe!std::_Atomic_uint::load(std::memory_order _Order=memory_order_relaxed) 第 630 行	C++
 	test_bitcoin.exe!BCLog::Logger::WillLogCategory(BCLog::LogFlags category=RAND) 第 83 行	C++
 	test_bitcoin.exe!LogAcceptCategory(BCLog::LogFlags category=RAND) 第 117 行	C++
 	test_bitcoin.exe!LogPrint<char [15],char [19],unsigned long>(const BCLog::LogFlags & category=RAND, const char[15] & <args_0>=..., const char[19] & <args_1>=..., const unsigned long & <args_2>=707688) 第 150 行	C++
>	test_bitcoin.exe!RandAddSeedPerfmon(CSHA512 & hasher={...}) 第 199 行	C++
 	test_bitcoin.exe!SeedStartup(CSHA512 & hasher={...}, `anonymous-namespace'::RNGState & rng={...}) 第 456 行	C++
 	test_bitcoin.exe!ProcRand(unsigned char * out=0x0000009f8cd8eb78, int num=8, RNGLevel level=FAST) 第 493 行	C++
 	test_bitcoin.exe!GetRandBytes(unsigned char * buf=0x0000009f8cd8eb78, int num=8) 第 508 行	C++
 	test_bitcoin.exe!GetRand(unsigned __int64 nMax=18446744073709551615) 第 523 行	C++
 	test_bitcoin.exe!SaltedTxidHasher::SaltedTxidHasher() 第 1093 行	C++
 	test_bitcoin.exe!boost::tuples::cons<SaltedTxidHasher,boost::tuples::cons<std::equal_to<uint256>,boost::tuples::null_type> >::cons<SaltedTxidHasher,boost::tuples::cons<std::equal_to<uint256>,boost::tuples::null_type> >() 第 284 行	C++
 	test_bitcoin.exe!boost::tuples::cons<mempoolentry_txid,boost::tuples::cons<SaltedTxidHasher,boost::tuples::cons<std::equal_to<uint256>,boost::tuples::null_type> > >::cons<mempoolentry_txid,boost::tuples::cons<SaltedTxidHasher,boost::tuples::cons<std::equal_to<uint256>,boost::tuples::null_type> > >() 第 284 行	C++
 	test_bitcoin.exe!boost::tuples::cons<unsigned __int64,boost::tuples::cons<mempoolentry_txid,boost::tuples::cons<SaltedTxidHasher,boost::tuples::cons<std::equal_to<uint256>,boost::tuples::null_type> > > >::cons<unsigned __int64,boost::tuples::cons<mempoolentry_txid,boost::tuples::cons<SaltedTxidHasher,boost::tuples::cons<std::equal_to<uint256>,boost::tuples::null_type> > > >() 第 284 行	C++
 	test_bitcoin.exe!boost::tuples::tuple<unsigned __int64,mempoolentry_txid,SaltedTxidHasher,std::equal_to<uint256>,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>::tuple<unsigned __int64,mempoolentry_txid,SaltedTxidHasher,std::equal_to<uint256>,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>() 第 485 行	C++
 	test_bitcoin.exe!boost::tuples::cons<boost::tuples::tuple<unsigned __int64,mempoolentry_txid,SaltedTxidHasher,std::equal_to<uint256>,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>,boost::tuples::cons<boost::tuples::tuple<boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByDescendantScore,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>,boost::tuples::cons<boost::tuples::tuple<boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByEntryTime,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>,boost::tuples::cons<boost::tuples::tuple<boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByAncestorFee,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>,boost::tuples::null_type> > > >::cons<boost::tuples::tuple<unsigned __int64,mempoolentry_txid,SaltedTxidHasher,std::equal_to<uint256>,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>,boost::tuples::cons<boost::tuples::tuple<boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByDescendantScore,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>,boost::tuples::cons<boost::tuples::tuple<boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByEntryTime,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>,boost::tuples::cons<boost::tuples::tuple<boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByAncestorFee,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type,boost::tuples::null_type>,boost::tuples::null_type> > > >() 第 284 行	C++
 	test_bitcoin.exe!boost::multi_index::multi_index_container<CTxMemPoolEntry,boost::multi_index::indexed_by<boost::multi_index::hashed_unique<mempoolentry_txid,SaltedTxidHasher,boost::mpl::na,boost::mpl::na>,boost::multi_index::ordered_non_unique<boost::multi_index::tag<descendant_score,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByDescendantScore>,boost::multi_index::ordered_non_unique<boost::multi_index::tag<entry_time,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByEntryTime>,boost::multi_index::ordered_non_unique<boost::multi_index::tag<ancestor_score,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByAncestorFee>,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,std::allocator<CTxMemPoolEntry> >::multi_index_container<CTxMemPoolEntry,boost::multi_index::indexed_by<boost::multi_index::hashed_unique<mempoolentry_txid,SaltedTxidHasher,boost::mpl::na,boost::mpl::na>,boost::multi_index::ordered_non_unique<boost::multi_index::tag<descendant_score,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByDescendantScore>,boost::multi_index::ordered_non_unique<boost::multi_index::tag<entry_time,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByEntryTime>,boost::multi_index::ordered_non_unique<boost::multi_index::tag<ancestor_score,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::identity<CTxMemPoolEntry>,CompareTxMemPoolEntryByAncestorFee>,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,std::allocator<CTxMemPoolEntry> >() 第 177 行	C++
 	test_bitcoin.exe!CTxMemPool::CTxMemPool(CBlockPolicyEstimator * estimator=0x00007ff689f81fb0) 第 328 行	C++
 	test_bitcoin.exe!`dynamic initializer for 'mempool''() 第 244 行	C++
 	test_bitcoin.exe!_initterm(void(*)() * first=0x00007ff68945e000, void(*)() * last=0x00007ff689461830) 第 22 行	C++
 	[外部程式碼]	
```
g_logger has not been newed.
https://github.com/bitcoin/bitcoin/blob/f080c65a09a8f3b223c9b5d8e3562320bf258fcd/src/logging.cpp#L24

sipa committed:
Don't log RandAddSeedPerfmon details

These are hard to deal with, as in a follow-up this function can get
called before the logging infrastructure is initialized.

sipa commented:
@ken2812221 Thanks so much; this was an actual bug.

jesskie commented:
Tnx

gmaxwell commented on src/random.cpp:
```diff
@@ -129,18 +135,41 @@ static bool GetHWRand(unsigned char* ent32) {
+        // Write benchmark data (into output hasher)
+        int64_t perf = GetPerformanceCounter();
+        hasher.Write((const unsigned char*)&perf, sizeof(perf));
+    } while (GetTimeMicros() < start + microseconds);
```
Maybe we should be checking for shutdown interruption in this loop?  At 10ms it isn't critical...

sipa replied:
Marking as resolved as the hardening code is removed from this PR.

jesskie commented:
What is d best thing to do?

gmaxwell commented on src/random.cpp:
```diff
@@ -272,86 +294,217 @@ void GetOSRand(unsigned char *ent32)
+    if (current_time > last_strengthen + 60000) {
+        // Write the actual RNG state into the hasher, so it goes through the strengthening as well.
+        unsigned char buf[32];
+        if (rng.MixExtract(buf, 32, CSHA512(), false)) {
```
I think this one can ignore the return value of mix extract.

sipa replied:
Done.

gmaxwell commented:
utACK

sipa commented:
Anything I can do to simplify review here? I can move the strengthening loop to a separate PR if that helps.

sipa committed:
DRY: Implement GetRand using FastRandomContext::randrange

sipa commented:
I've removed the hash strengthening from this PR; I'll open a separate PR for it later.

practicalswift commented:
Concept ACK

Bonus points for the removal of `NO_THREAD_SAFETY_ANALYSIS` (for `locking_callback`) and also the various `noexcept` additions.

laanwj commented:
> Bonus points for the removal of NO_THREAD_SAFETY_ANALYSIS (for locking_callback) and also the various noexcept additions.

Getting rid of `CInit` is wonderful too :tada:

utACK 3a82a46eaa1f1c9e55635f8940a2bda7d3d2a817

instagibbs commented:
@sipa when you say it implements "most of" the issue, which parts are missing?

sipa commented:
@instagibbs There are some potential entropy sources that can still be added (environment data, and a strengthening loop). And of course removing OpenSSL once we have sufficient replacements for all of it.

instagibbs commented:
"Abstract out seeding/extracting entropy into RandExtract." <--- commit message refers to presumably old name of function

instagibbs reviewed (approved):
utACK https://github.com/bitcoin/bitcoin/pull/14955/commits/3a82a46eaa1f1c9e55635f8940a2bda7d3d2a817, though I'd like a "RAND_add" explanation as noted

sipa commented on src/random.cpp:
```diff
@@ -329,61 +319,128 @@ RNGState& GetRNGState()
+    if (level != RNGLevel::FAST) {
+        unsigned char buf[64];
+        CSHA512().Write(out, num).Finalize(buf);
+        RAND_add(buf, sizeof(buf), 8 * num);
```
I misunderstood the API and assumed the entropy argument was measured in bits.

I don't think this matters much, modern RNGs don't really measure entropy anymore. But I'll fix it.

sipa replied:
Fixed.

sipa commented on src/random.cpp:
```diff
@@ -292,6 +292,32 @@ struct RNGState {
     }
+
+    /** Extract up to 32 bytes of entropy from the RNG state, mixing in new entropy from hasher. */
+    void MixExtract(unsigned char* out, int num, CSHA512&& hasher)
```
Fixed.

sipa commented on src/random.cpp:
```diff
@@ -292,6 +292,32 @@ struct RNGState {
+    /** Extract up to 32 bytes of entropy from the RNG state, mixing in new entropy from hasher. */
+    void MixExtract(unsigned char* out, int num, CSHA512&& hasher)
+    {
+        assert(num >= 0 && num <= 32);
```
Done (by adding an assert in the `if (num) {` branch).

instagibbs commented:
re-utACK https://github.com/bitcoin/bitcoin/pull/14955/commits/b1ee8608f2b1700505e24f09ee141d4608c36078

sipa committed:
Encapsulate RNGState better

sipa committed:
Use secure allocator for RNG state

sipa commented:
Added two commits:

* Turn RNGState into a class with private fields (makes it easier to make sure no code can accidentally wipe it or so), suggested by @pstratem on IRC.
* Allocate the RNGState in the mlocked pool, suggested by @gmaxwell on IRC.

ryanofsky reviewed (commented):
Mostly reviewed this, but need to finish up last few commits. Two suggestions:

- It would be good to add information from #14623 and #14955 descriptions to a code comment (or maybe a doc/random.md document) to document the new design. It would be a shame if this PR were merged as-is and all the considerations that went into designing it wound up buried in commit history.

- There should be a release note about this change. Probably not necessary to go into detail about the design, but it would be good to have a short overall description and call out any changes to user-visible behavior like logging.

My review status:

- [X] 5dfb99059491b1db9133cf0e6b0158ff4bdc9ef0 Use FRC::randbytes instead of reading >32 bytes from RNG (1/13)
- [X] ede7e302d55eb37936268caffcc4b31839e58283 Don't log RandAddSeedPerfmon details (2/13)
- [X] 2ce2c827115748ebea121e6e2abfded260f50ea6 Remove adding timestamps in sanity check to OpenSSL RNG (3/13)
- [X] 2dcb3e938a210b1bbaa178ad6f9a5d418c4e2939 Automatically initialize RNG on first use. (4/13)
- [X] d29f3b967313db206b43c92df93f76cafa0fb843 Add thread safety annotations to RNG state (5/13)
- [X] cef6d52cbed7f30bcdec5b6c5a72e9a49e26eeab Abstract out seeding/extracting entropy into RNGState::MixExtract (6/13)
- [ ] 7c82fd09b9e7bdf62f8aa4d4d92acf9865a0476b Switch all RNG code to the built-in PRNG. (7/13)
- [ ] d121e29d26836efaad600d5cfb1fcc884c48d17a Remove hwrand_initialized. (8/13)
- [ ] 5749bfea805a1c37b2122ebc42827d4fef1fc25b Sprinkle some sweet noexcepts over the RNG code (9/13)
- [ ] 214fcf143c3f2cafb8d89c551d5de00b5da6ffa5 Use our own locking instead of using OpenSSL in multithreaded mode (10/13)
- [ ] b1ee8608f2b1700505e24f09ee141d4608c36078 DRY: Implement GetRand using FastRandomContext::randrange (11/13)
- [ ] 0259c0bbdd8cb4dd6d924c78279bb9382fbdc445 Encapsulate RNGState better (12/13)
- [ ] 10a00393aecafe42cef65074f04656f2f05e9d13 Use secure allocator for RNG state (13/13)

sipa committed:
Document RNG design in random.h

sipa commented:
@ryanofsky Thanks for the review. I've addressed most of your comments, and added a bunch of comments to random.h. I'll let others comment on the need for adding release notes about it, but my thinking was to only make a note once we fully remove OpenSSL (which may or may not be in the same release).

I've tried to not interfere with existing review too much by keeping the commit order and not rebased, but if desired by reviewers I can probably split up the big "Switch all RNG code to the built-in PRNG." commit.

sipa commented on src/random.cpp:
```diff
@@ -292,6 +292,33 @@ struct RNGState {
+    void MixExtract(unsigned char* out, size_t num, CSHA512&& hasher)
+    {
+        assert(num <= 32);
+        unsigned char buf[64];
```
I had to change it to `CSHA512::OUTPUT_SIZE`, as clang didn't agree `hasher.OUTPUT_SIZE` was a compile-time constant.

laanwj commented:
> I'll let others comment on the need for adding release notes about it, but my thinking was to only make a note once we fully remove OpenSSL (which may or may not be in the same release).

IMO, a release note is unnecessary for this (apart from the automatically-generated changelog line). It is an internal change, there are hardly or no user-visible changes, the user doesn't need to *do* anything like change the configuration.

ryanofsky reviewed (approved):
utACK 76831bf79bb453dc00bc8f5a94a784581e1212d3. Changes since last review: dropped SanityCheck cleanup commit 2ce2c827115748ebea121e6e2abfded260f50ea6, added documentation commit 76831bf79bb453dc00bc8f5a94a784581e1212d3, made various suggested changes in other commits.

- [X] 6a57ca91da23c6a5d91399ffc7fc09a99b6d4c76 Use FRC::randbytes instead of reading >32 bytes from RNG (1/13)
- [X] 1a3b26e8e3ee64420cb5081de97ce1944f8b9a7a Don't log RandAddSeedPerfmon details (2/13)
- [X] 2a02d2c369ee06cbc334a0dd7150ec72ef83a182 Automatically initialize RNG on first use. (3/13)
- [X] 9a4aae5427a3a71298413b9348073512dbb43056 Add thread safety annotations to RNG state (4/13)
- [X] d76cc6dbb9e1c5c6a0f3b002932ba4033b722255 Abstract out seeding/extracting entropy into RNGState::MixExtract (5/13)
- [X] 85f8f99e0ae9d1c98f676ae8722cda4b967dcfb4 Switch all RNG code to the built-in PRNG. (6/13)
- [X] 604ae2bb9789a213711f5f53e0c74225e74bb085 Remove hwrand_initialized. (7/13)
- [X] 6ff775ff80b090494d115753ddcf488a3cf923a9 Sprinkle some sweet noexcepts over the RNG code (8/13)
- [X] fa2eabe154d00bf6c54f604acbe758186d044639 Use our own locking instead of using OpenSSL in multithreaded mode (9/13)
- [X] d8972a68a63a2dddd910d340b8ff05b81fd877bc DRY: Implement GetRand using FastRandomContext::randrange (10/13)
- [X] 7939daa351a31a36f2086771611b501418536989 Encapsulate RNGState better (11/13)
- [X] 9505dc55c4fcaa61d0fc97558dd22c7fdb879900 Use secure allocator for RNG state (12/13)
- [x] 76831bf79bb453dc00bc8f5a94a784581e1212d3 Document RNG design in random.h (13/13)

sipa committed:
Integrate util/system's CInit into RNGState

This guarantees that OpenSSL is initialized properly whenever randomness
is used, even when that randomness is invoked from global constructors.

Note that this patch uses Mutex directly, rather than CCriticalSection.
This is because the lock-detection code is not necessarily initialized
during global constructors.

sipa commented on src/random.cpp:
```diff
@@ -298,6 +298,34 @@ struct RNGState {
+        }
+        // Best effort cleanup of internal state
+        hasher.Reset();
+        memory_cleanse(buf, 64);
```
Perhaps it does, though we're all over the code only using memory_cleanse for memory buffers directly, and I feel slightly uneasy to invoke on the state of complex objects.

Maybe in a separate change we can add a Cleanse() method to CSHA512 and others, and start using those were useful?

sipa commented on src/random.cpp:
```diff
@@ -174,7 +164,7 @@ static void RandAddSeedPerfmon()
     }
     RegCloseKey(HKEY_PERFORMANCE_DATA);
     if (ret == ERROR_SUCCESS) {
-        RAND_add(vData.data(), nSize, nSize / 100.0);
```
Done.

sipa commented on src/random.cpp:
```diff
@@ -337,61 +327,128 @@ RNGState& GetRNGState()
+    hasher.Write((const unsigned char*)&perfcounter, sizeof(perfcounter));
+}
 
-void RandAddSeedSleep()
```
Done.

sipa commented on src/random.cpp:
```diff
@@ -337,61 +327,128 @@ RNGState& GetRNGState()
+    SeedTimestamp(hasher);
 }
 
-void GetStrongRandBytes(unsigned char* out, int num)
```
Done.

sipa commented on src/random.cpp:
```diff
@@ -441,11 +441,11 @@ static void ProcRand(unsigned char* out, int num, RNGLevel level)
-void GetRandBytes(unsigned char* buf, int num) { ProcRand(buf, num, RNGLevel::FAST); }
-void GetStrongRandBytes(unsigned char* buf, int num) { ProcRand(buf, num, RNGLevel::SLOW); }
+void GetRandBytes(unsigned char* buf, int num) noexcept { ProcRand(buf, num, RNGLevel::FAST); }
+void GetStrongRandBytes(unsigned char* buf, int num) noexcept { ProcRand(buf, num, RNGLevel::SLOW); }
```
I've added a comment on the use of noexcept and why it's not used for everything.

sipa commented on src/util/system.cpp:
```diff
@@ -86,47 +83,6 @@ const char * const BITCOIN_PID_FILENAME = "bitcoind.pid";
-    ~CInit()
-    {
-        // Securely erase the memory used by the PRNG
-        RAND_cleanup();
```
I've reverted this change for other reasons. I don't think RAND_cleanup is particularly important, but let's keep it for now as long as we're using the OpenSSL RNG.

sipa commented on src/util/system.cpp:
```diff
@@ -86,47 +83,6 @@ const char * const BITCOIN_PID_FILENAME = "bitcoind.pid";
-    {
-        // Init OpenSSL library multithreading support
-        ppmutexOpenSSL.reset(new CCriticalSection[CRYPTO_num_locks()]);
-        CRYPTO_set_locking_callback(locking_callback);
```
Right, that was the reasoning. I assumed the payment server code was already not multithreaded (because it's synchronized through the UI), but reading more about OpenSSL's multithreading design, it seems that X509 processing has a bunch of global state which needs locking, so your comment made me uneasy about this.

I've replaced it with a (less satisfying...) commit that moves the `CInit` logic to `RNGState`. This is necessary, as we're now invoking OpenSSL based RNG code from global constructors, possibly before the CInit object is creator (and RNGState is always created whenever needed, even before global constructors). Perhaps this was a problem in the old code too.

Sjors commented on src/random.cpp:
```diff
@@ -380,61 +362,128 @@ void LockingCallbackOpenSSL(int mode, int i, const char* file, int line) NO_THRE
-    // Combine with and update state
-    AddDataToRng(&nPerfCounter1, sizeof(nPerfCounter1), rng);
-    AddDataToRng(&nPerfCounter2, sizeof(nPerfCounter2), rng);
+    // Hardware randomness is very fast when available; use it always.
```
Assuming it's always CPU based and not same external device, so this assumption should be stated in the empty / placeholder `HWRandInit` implementations for other architectures.

sipa replied:
Done.

Sjors reviewed (approved):
tACK 3b001b0 on macOS 10.14.2

Despite the volume of my comments, they're all about additional clarification, which can always be done later (though I think they'll be helpful for other reviewers).

This new code is more readable than before, which is very nice for something as critical as RNG stuff. The breakdown into smaller commits is really helpful.

This needs testing on Windows, e.g. because there's `#ifdef WIN32` stuff.

Checking the benchmarks would also be useful (IIRC there's a bot for that).

* 6a57ca91da23c6a5d91399ffc7fc09a99b6d4c76 "There was only one place in the codebase where we're directly reading >32 bytes from the RNG": what happens in that case? If bad, maybe add an assert or comment in `GenerateAuthCookie` that `const size_t COOKIE_SIZE = 32;` shouldn't be increased.

* 1a3b26e8e3ee64420cb5081de97ce1944f8b9a7a the usage of `RandAddSeedPerfmon` is a bit weird

It's called by all operating systems, but its entire content is wrapped in `#ifdef WIN32`. On top of that, it can now fail without complaining (because it's non-critical). If `static bool warned` is hard to deal with, why not give the function a return value?  Maybe move both the `ifdef` and failure handling to the caller of this function (with or without the log statement). Both callers already know this is a Windows only thing, as evidenced by their comments.

* 2a02d2c369ee06cbc334a0dd7150ec72ef83a182 I got confused by the names  `HWRandInit` and `HWRandReport`, so suggested alternative names inline.

* 772fce745c51aa9d09f93e13eeb822e889ae25f7  OpenSSL [docs](https://www.openssl.org/docs/man1.0.2/crypto/threads.html) say "OpenSSL can generally be used safely in multi-threaded applications provided that at least two callback functions are set, the locking_function and threadid_func", but we're not using the latter. Do we meet the conditions stated further down the doc for when that's safe?

* b83c06a622652853238f3c35b9303194eafec1c1 what benefits do you expect from the `noexcept` sprinkling? Safety, readability or performance? What's the worst case scenario if you're wrong about any of them?

* 63dc0adf7c98b20eb9ca590c97547b9681d2b48c previously we only used `secure_allocator` in `CKey` and (wallet) `CCrypter`. Not sure if it matters at all to expand its usage, but I don't pretent to even remotely understand the existing magic in `secure_allocator` (written by @theuni four years ago).

sipa commented:
> 6a57ca9 "There was only one place in the codebase where we're directly reading >32 bytes from the RNG": what happens in that case? If bad, maybe add an assert or comment in GenerateAuthCookie that const size_t COOKIE_SIZE = 32; shouldn't be increased.

There already is an assert: `assert(num <= 32);` in `ProcRand()`.

> 1a3b26e the usage of RandAddSeedPerfmon is a bit weird. It's called by all operating systems, but its entire content is wrapped in #ifdef WIN32. On top of that, it can now fail without complaining (because it's non-critical). If static bool warned is hard to deal with, why not give the function a return value? Maybe move both the ifdef and failure handling to the caller of this function (with or without the log statement). Both callers already know this is a Windows only thing, as evidenced by their comments.

As I've pointed out before, we may want to add similar perfmon data for other platforms.

> 772fce7 OpenSSL docs say "OpenSSL can generally be used safely in multi-threaded applications provided that at least two callback functions are set, the locking_function and threadid_func", but we're not using the latter. Do we meet the conditions stated further down the doc for when that's safe?

The docs also point out that on systems with thread-safe errno, its address is used as thread identifier. I believe that's the case on all supported platforms.

> b83c06a what benefits do you expect from the noexcept sprinkling? Safety, readability or performance? What's the worst case scenario if you're wrong about any of them?

Making the function's behavior more explicit to developers, and performance. The worst case about being wrong about them is that certain optimizations can't be used (for example, when a constructor can directly or indirectly throw, more complicated deconstructors are needed; or when a move operator can throw it can't be used for efficiently moving data between containers and necessitating a copy/delete instead).

> 63dc0ad previously we only used secure_allocator in CKey and (wallet) CCrypter. Not sure if it matters at all to expand its usage, but I don't pretent to even remotely understand the existing magic in secure_allocator

It shouldn't. It's the right tool for the job (preventing sensitive material from leaking into swap files), so I don't see why we shouldn't use it.

sipa commented on src/random.h:
```diff
@@ -13,33 +13,81 @@
+ *   - A high-precision timestamp (rdtsc when available, c++ high_resolution_clock otherwise)
+ *   - Hardware RNG (rdrand) when available.
+ *   These entropy sources are very fast, and mostly designed to protect against situations
+ *   where a VM state restore/copy results in multiple systems with the same randomness.
```
> Does this refer to an accidental scenario, where a user ends up generating the same private keys or accidentally signing with the same nonce? Or also a malicious scenario?

Does it matter whether it can be triggered accidentally or intentionally?

> So do I understand correctly that the reason FastRandomContext needs to fast is (mostly) this, or are there also areas of code that need the additional performance?

I don't understand the question or the relation with the code you're commenting on.

Sjors replied:
I was trying to understand _why_ it needs to be fast. The explanation here suggests it's mostly about protecting against some sort of VM replay scenario. But in later commits it's clear that there's also a performance concern, e.g. the Perfmon can take two seconds.

sipa replied:
Ah, I see. I think you're looking at it the wrong way. Of course there are many places in the code where we need fast randomness (of various sizes); the issue that the speed concern means we can't do very good seeding, so we need a separate function for the few places where we _really_ need good randomness (GetStrongRandBytes).

Sjors replied:
I do understand that fast also implies weak. The question is why you would ever accept weak, if you already code from strong. The obvious answer is performance. But the comment here talks about a VM replay scenario.

But now I see what confused me: "These entropy sources are very fast, and mostly designed to protect against situations  where a VM state restore" - the part after "and" doesn't refer to "fast"; they are an _additional_ benefit of these sources.

Sjors replied:
FastButStillHardToReplayRandomContext :-P

sipa replied:
No, `FastRandomContext` does *not* protect against replay. `GetRand*()` does.

sipa commented on src/random.h:
```diff
@@ -178,7 +178,12 @@ void GetOSRand(unsigned char *ent32);
+ * Initialize global RNG state and log any CPU features that are used.
+ *
+ * Calling this function is optional. RNG state will be initialized when first
+ * needed if it is not called.
```
No, it does refer to the broader sense, including seeding.

sipa commented on src/random.cpp:
```diff
@@ -298,6 +298,34 @@ struct RNGState {
+        unsigned char buf[64];
+        static_assert(sizeof(buf) == CSHA512::OUTPUT_SIZE, "Buffer needs to have hasher's output size");
+        {
+            LOCK(cs);
```
The only difference is that it lets you specify the name of the lock variable.

Sjors commented on src/random.cpp:
```diff
@@ -166,13 +166,12 @@ static void RandAddSeedPerfmon()
-            LogPrintf("%s: Warning: RegQueryValueExA(HKEY_PERFORMANCE_DATA) failed with code %i\n", __func__, ret);
-            warned = true;
-        }
+        // Performance data is only used as a last resort to improve entropy.
```
The way I read the phrase "as a last resort" is: *if* there was a way to detect that our RNG attempts thus far have failed, we try this thing as a last resort. In the last resort also fails, that's critical. Hence the contraction.

But because we normally can't detect if we actually needed the last resort, we don't know if the failure was critical, and most likely it wasn't.

sipa replied:
Reformulated the text a bit.

Sjors commented on src/random.cpp:
```diff
@@ -74,22 +74,32 @@ static inline int64_t GetPerformanceCounter()
     hwrand_initialized.store(true);
 }
+
+static void HWRandReport()
```
I spent at least half an hour being confused about what `HWRandInit` and `HWRandReport` were doing, otherwise I wouldn't have brought up this up.

ryanofsky commented on src/random.cpp:
```diff
@@ -336,6 +337,8 @@ struct RNGState {
         return ret;
     }
+
+    Mutex& GetOpenSSLLock(int i) { return cs_openssl[i]; }
```
In commit "Encapsulate RNGState better" (bfd3a23f277c16236ba6ea7391a98277226c1470)

Seems more accurate to call it Mutex than Lock (as in 7939daa351a31a36f2086771611b501418536989), if Mutex is the object that gets locked, and Lock is the object that locks it.

sipa replied:
Done.

ryanofsky reviewed (approved):
utACK 3b001b09e840c04b06c0ec6689ed666d68e58cb1. Changes since last review: moving openssl init/locking code instead of removing it, adding commit comments about openssl seeding and function renames, adding code comment about noexcept.

sipa committed:
Rename some hardware RNG related functions

sipa commented on src/random.cpp:
```diff
@@ -166,13 +166,12 @@ static void RandAddSeedPerfmon()
-            warned = true;
-        }
+        // Performance data is only used as a last resort to improve entropy.
+        // Failure to read it is isn't conidered critical, so we don't call
```
Fixed.

sipa commented on src/random.cpp:
```diff
@@ -74,22 +74,32 @@ static inline int64_t GetPerformanceCounter()
 static bool rdrand_supported = false;
 static constexpr uint32_t CPUID_F1_ECX_RDRAND = 0x40000000;
-static void RDRandInit()
+static void HWRandInit()
```
Done, added a separate commit with renames and the comments on HWRNG performance you suggested elsewhere.

sipa commented on src/random.cpp:
```diff
@@ -74,22 +74,32 @@ static inline int64_t GetPerformanceCounter()
+
 #else
-static void RDRandInit() {}
+static void HWRandInit() {}
```
Added a comment.

sipa commented on src/random.cpp:
```diff
@@ -278,6 +288,26 @@ void GetRandBytes(unsigned char* buf, int num)
+namespace {
+struct RNGState {
+    Mutex cs;
+    unsigned char state[32] = {0};
```
Added a similar comment to the RNGState::m_state variable.

sipa commented on src/random.cpp:
```diff
@@ -297,6 +280,7 @@ struct RNGState {
     Mutex cs;
     unsigned char state[32] GUARDED_BY(cs) = {0};
     uint64_t counter GUARDED_BY(cs) = 0;
+    bool initialized GUARDED_BY(cs) = false;
```
Renamed to `m_strongly_seeded`.

sipa commented on src/random.cpp:
```diff
@@ -380,61 +362,128 @@ void LockingCallbackOpenSSL(int mode, int i, const char* file, int line) NO_THRE
-    // Third source: HW RNG, if available.
-    if (GetHWRand(buf)) {
-        hasher.Write(buf, 32);
+    // Windows performance monitor data (once every 10 minutes)
```
Oh, I see the confusion. Removed that phrasing.

sipa commented on src/random.cpp:
```diff
@@ -278,6 +288,26 @@ void GetRandBytes(unsigned char* buf, int num)
 
+namespace {
+struct RNGState {
+    Mutex cs;
```
Done.

Sjors commented:
Wonderful, re-tACK 223de8d

sipa commented:
FWIW, I've benchmarked this before and after this PR on my system. GetRandBytes goes from around ~1 μs to ~3 μs. GetStrongRandBytes stays around ~10 μs. Note that GetRandBytes is no longer called inside tight loops since #14624.

ryanofsky reviewed (approved):
utACK 223de8d94d6522f795ec3c2e7db27469f24aa68c. All changes since last review are renames or comment improvements.
